File: lib/backup_stack.py
"""AWS Backup for Aurora database."""

from constructs import Construct
from cdktf_cdktf_provider_aws.backup_vault import BackupVault
from cdktf_cdktf_provider_aws.backup_plan import BackupPlan, BackupPlanRule, BackupPlanRuleLifecycle
from cdktf_cdktf_provider_aws.backup_selection import BackupSelection, BackupSelectionSelectionTag
from cdktf_cdktf_provider_aws.iam_role import IamRole
from cdktf_cdktf_provider_aws.iam_role_policy_attachment import IamRolePolicyAttachment


class BackupStack(Construct):
    """AWS Backup for Aurora database."""

    def __init__(self, scope: Construct, construct_id: str, environment_suffix: str,
                 region: str, provider, aurora_cluster_arn: str):
        super().__init__(scope, construct_id)

        # Backup vault
        vault = BackupVault(
            self, "vault",
            name=f"payment-backup-vault-{environment_suffix}",
            tags={"Name": f"payment-backup-vault-{environment_suffix}"},
            provider=provider,
        )

        # IAM role
        backup_role = IamRole(
            self, "backup_role",
            name=f"payment-backup-role-{environment_suffix}",
            assume_role_policy='{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"backup.amazonaws.com"},"Action":"sts:AssumeRole"}]}',
            tags={"Name": f"payment-backup-role-{environment_suffix}"},
            provider=provider,
        )

        for policy in ["service-role/AWSBackupServiceRolePolicyForBackup", "service-role/AWSBackupServiceRolePolicyForRestores"]:
            IamRolePolicyAttachment(
                self, f"attach_{policy.split('/')[-1]}",
                role=backup_role.name,
                policy_arn=f"arn:aws:iam::aws:policy/{policy}",
                provider=provider,
            )

        # Backup plan
        backup_plan = BackupPlan(
            self, "backup_plan",
            name=f"payment-backup-plan-{environment_suffix}",
            rule=[BackupPlanRule(
                rule_name="daily_backup",
                target_vault_name=vault.name,
                schedule="cron(0 3 * * ? *)",
                lifecycle=BackupPlanRuleLifecycle(delete_after=7),
            )],
            tags={"Name": f"payment-backup-plan-{environment_suffix}"},
            provider=provider,
        )

        # Backup selection
        BackupSelection(
            self, "backup_selection",
            name=f"payment-aurora-backup-{environment_suffix}",
            plan_id=backup_plan.id,
            iam_role_arn=backup_role.arn,
            resources=[aurora_cluster_arn],
            selection_tag=[BackupSelectionSelectionTag(
                type="STRINGEQUALS",
                key="Name",
                value=f"payment-cluster-{environment_suffix}",
            )],
            provider=provider,
        )

File: archive/pulumi-ts/Pr7149/lib/backup-stack.ts
import * as aws from '@pulumi/aws';
import * as pulumi from '@pulumi/pulumi';

export interface BackupStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
  clusterArn: pulumi.Input<string>;
}

export class BackupStack extends pulumi.ComponentResource {
  constructor(
    name: string,
    args: BackupStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:backup:BackupStack', name, args, opts);

    const { environmentSuffix, tags, clusterArn } = args;

    // Backup vault
    const backupVault = new aws.backup.Vault(
      `payment-backup-vault-${environmentSuffix}`,
      {
        name: `payment-backup-vault-${environmentSuffix}`,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-backup-vault-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // DR vault in us-west-2 (must be created before backup plan to reference its ARN)
    const drProvider = new aws.Provider(
      `dr-provider-${environmentSuffix}`,
      {
        region: 'us-west-2',
      },
      { parent: this }
    );

    const drBackupVault = new aws.backup.Vault(
      `payment-backup-vault-${environmentSuffix}-dr`,
      {
        name: `payment-backup-vault-${environmentSuffix}-dr`,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-backup-vault-${environmentSuffix}-dr`,
          Region: 'us-west-2',
        })),
      },
      { parent: this, provider: drProvider }
    );

    // Backup plan
    const backupPlan = new aws.backup.Plan(
      `payment-backup-plan-${environmentSuffix}`,
      {
        name: `payment-backup-plan-${environmentSuffix}`,
        rules: [
          {
            ruleName: 'DailyBackup',
            targetVaultName: backupVault.name,
            schedule: 'cron(0 2 * * ? *)',
            lifecycle: {
              deleteAfter: 30,
            },
            copyActions: [
              {
                destinationVaultArn: drBackupVault.arn,
                lifecycle: {
                  deleteAfter: 30,
                },
              },
            ],
          },
        ],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-backup-plan-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Backup role
    const backupRole = new aws.iam.Role(
      `payment-backup-role-${environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: {
                Service: 'backup.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            },
          ],
        }),
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-backup-role-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Attach AWS managed backup policy
    new aws.iam.RolePolicyAttachment(
      `payment-backup-policy-${environmentSuffix}`,
      {
        role: backupRole.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup',
      },
      { parent: this }
    );

    new aws.iam.RolePolicyAttachment(
      `payment-backup-restore-policy-${environmentSuffix}`,
      {
        role: backupRole.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForRestores',
      },
      { parent: this }
    );

    // Backup selection
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const backupSelection = new aws.backup.Selection(
      `payment-backup-selection-${environmentSuffix}`,
      {
        name: `payment-rds-selection-${environmentSuffix}`,
        planId: backupPlan.id,
        iamRoleArn: backupRole.arn,
        resources: [clusterArn],
      },
      { parent: this }
    );

    this.registerOutputs({
      backupVaultName: backupVault.name,
      backupPlanId: backupPlan.id,
      drBackupVaultName: drBackupVault.name,
    });
  }
}
